pkg/gitserver/controller/controller.go

```diff
@@ -0,0 +1,231 @@
+
+	// Get the repository name if the BuildConfig points to a repository on this git server
+	repoName, hasGitServerSource := c.gitServerRepository(bc)
+
+	if !hasGitServerSource {
+		// If the BuildConfig's source doesn't point to this git server,
+		// ensure that any references to it are removed from existing git repositories.
+		entry, inCache := c.BCCache[bc.Name]
+		if inCache {
+			if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+				return err
+			}
+			delete(c.BCCache, bc.Name)
+		}
+		return nil
+	}
+
+	removed := false
+	entry, inCache := c.BCCache[bc.Name]
+	if inCache && (entry.Repository != repoName || entry.Reference != bc.Spec.Source.Git.Ref) {
+		removed = true
+		// If already exists in cache, but either the repository or reference don't match anymore,
+		// remove the reference from the git repository
+		if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+			return err
+		}
+		delete(c.BCCache, bc.Name)
+	}
+
+	if !inCache || removed {
+		// If not already in the cache or it has been removed because of a mismatch, add to the
+		// repository config
+		if err := c.addToRepository(repoName, bcEntry(bc.Name, bc.Spec.Source.Git.Ref)); err != nil {
+			return err
+		}
+		c.BCCache[bc.Name] = BCCacheEntry{
+			Repository: repoName,
+			Reference:  bc.Spec.Source.Git.Ref,
+		}
+	}
+
+	return nil
+}
+
+func (c *BuildConfigController) HandleBuildConfigDelete(bc *buildapi.BuildConfig) error {
+	// First remove from the cache (and corresponding repository) if present
+	entry, inCache := c.BCCache[bc.Name]
+	if inCache {
+		if err := c.removeFromRepository(entry.Repository, bcEntry(bc.Name, entry.Reference)); err != nil {
+			return err
+		}
+		delete(c.BCCache, bc.Name)
+	}
+	// Ensure that even if it's not in the cache, that its
+	// corresponding repository is updated if there is one.
+	repoName, hasGitServerSource := c.gitServerRepository(bc)
+	if hasGitServerSource {
+		if !inCache || entry.Repository != repoName || entry.Reference != bc.Spec.Source.Git.Ref {
+			if err := c.removeFromRepository(repoName, bcEntry(bc.Name, bc.Spec.Source.Git.Ref)); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *BuildConfigController) gitServerRepository(bc *buildapi.BuildConfig) (name string, isRepo bool) {
+	if bc.Spec.Source.Git == nil {
+		return "", false
+	}
+	parsedURL, err := url.Parse(bc.Spec.Source.Git.URI)
+	if err != nil {
+		return "", false
+	}
+	if parsedURL.Host != c.PublicURL.Host ||
+		parsedURL.Scheme != c.PublicURL.Scheme {
+		return "", false
+	}
+	path := parsedURL.Path
+	if c.PublicURL.Path != "" {
+		if !strings.HasPrefix(path, c.PublicURL.Path) {
+			return "", false
+		}
+		path = strings.TrimPrefix(path, c.PublicURL.Path)
+	}
+	path = strings.Trim(path, "/")
+	parts := strings.Split(path, "/")
+	if len(parts) > 1 {
+		return "", false
+	}
+	if !strings.HasSuffix(path, ".git") {
```

Comment:
It's added for consistency. You can create a repo ending with .git or without and from the client's perspective it's going to be the same thing. However, the server-side code could potentially create different dirs (one with and one without). This ensures that we always create the one with the suffix. We then create a symlink to the one without the suffix so they end up being the same thing.